### ISSUE TITLE
Multi-GPU Training

### ISSUE THREAD

**Issue by yusufali98** [UMass-Embodied-AGI/Mirage#9]
Hi,

Thanks for the cool work !

Does the code currently support multi-GPU training ? I tried training with 8 GPUs using deepspeed zero3 but observed significantly higher success rates for both Stage 1 (+42%) and Stage 2 (+20%) training when comparing to `Avg Success` reported numbers in Fig. 4 of [paper](https://www.arxiv.org/pdf/2506.17218) which I dont think is the expected result. I kept all other parameters exactly same as suggested in the README commands. Also, if multi-GPU is not supported would you know why the success increases significantly ?

**Comment by Unicorn53547** [comment]
Hi, 

Thanks for your interest in our work!

The numbers in Fig. 4 of paper come from model fine-tuned with CoT data that hasn't been public yet. With the released direct-SFT data, we achieve about 66 % success rate in Stage 1 and 76 % in Stage 2—match what you’re observing, so your run looks correct.

As for Multi-GPU training,  the current code wasn’t built for such setting, though running on several GPUs shouldn’t change accuracy much. We do plan to release code with multi-GPU support in the future and will announce it once ready.

Hope that clears things up—let me know if anything else comes up!

**Comment by yusufali98** [comment]
Hi,
Thanks for confirming that the multi-GPU run success rate is as expected.

In the single GPU run, I observed success of 7.2% and 50.5% with Stage 1 and 2 respectively. Earlier I thought this was okay since it was kind of close to the numbers in Table 4 of paper (still with a difference though). But it seems these success rates are much lower when compared to the expected 66% (Stage 1) and 76% (Stage 2) success rates. Ideally we should have the same success rates for both single/multi-GPU runs.

So would you know why the single GPU run has much lower SR in comparison to multi-GPU run ?
- Maybe it is because the effective batch size is only 8 since the single-GPU run uses [per_device_train_batch_size=1](https://github.com/UMass-Embodied-AGI/Mirage/blob/main/src/main.py#L179) and `--gradient_accumulation_steps 8` (from README steps) ? But I think you also use the same effective batch size since all results in paper are with single GPU
- I also want to mention that I am running all experiments with the `Qwen2.5-VL-3B-Instruct` model but not sure that should affect things since we get expected numbers in multi-GPU runs

Really appreciate your help with this !

**Comment by Unicorn53547** [comment]
Hi,

The 0.66/0.76 acc com from 7b model. Could you verify if you could get consistent resulls on 7B model using single gpu? From our run, single / mult gpu setting produce similiar results on 7b model (~0.76 on stage2).

**Comment by yusufali98** [comment]
Hi,
Sorry I am unable to run training with 7B model due to computational constraints. I will stick to the 3B, multi-GPU setting for now since I was able to reproduce the numbers you mentioned for Stage 1 and Stage 2

I had one more quick follow-up: I was trying to understand the effects of Stage 1 and Stage 2 training by running ablations in Figure 4 of paper but with the direct-SFT data currently released (instead of using CoT-finetuned model as you mentioned). I observe that Stage 1 only and Stage1+Stage 2 accuracy is as expected 0.63/0.776 respectively. But when I train with Stage 2 only, I get 0.78 accuracy which seems pretty high. Do you know if this is an expected result ?

Thanks for your really quick responses !

**Comment by Unicorn53547** [comment]
Hi, sorry for the late response.

Yes, this is expected for the VSP planning (direct) setting—we’ve also seen Stage 2-only match or slightly exceed Stage 1+2 in our experiments. That said, it’s a rare case: for other tasks, and for VSP planning with thoughts, Stage 1+2 consistently outperforms Stage 2-only.

We conjecture a few possible reasons for this behavior:

* Uncertain format/utility of path-based helper images:
In tasks like VSP reasoning or spatial reasoning, the entire helper image represents a meaningful state;while  in VSP planning, the crucial information lies only in the annotated path (red line). As also shown in the zero-shot upper-bound experiments in Fig. 6, the performance gain is limited compared to other tasks, it remains unclear whether the model reliably interprets the red-line path, and whether simply overlaying a path on the original image is an optimal helper design.

* Suboptimal image pooling for path-based helpers:
For tasks where the whole image encodes state, average pooling preserves useful information and can be beneficial for reasoning. In VSP planning, the information is concentrated along the path, so simple global average pooling may fail to capture the signal  effectively and could even introduce noise without additional text or reasoning chains.

Our primary focus is on interleaved reasoning (i.e.  text-visual-text trajectories), where text provides the step-by-step reasoning and also explains (how to reach) visual cues to some extent, and visual signals further support the reasoning and lead to the final answer. In such cases, mismatch/misunderstanding of visual information is far less likely.

We are currently experimenting with compression strategies as a future work. We are also working to release more data—stay tuned.

Hope this clarifies some confusion!

**Comment by llltttwww** [comment]
> Hi,
> 
> Thanks for the cool work !
> 
> Does the code currently support multi-GPU training ? I tried training with 8 GPUs using deepspeed zero3 but observed significantly higher success rates for both Stage 1 (+42%) and Stage 2 (+20%) training when comparing to `Avg Success` reported numbers in Fig. 4 of [paper](https://www.arxiv.org/pdf/2506.17218) which I dont think is the expected result. I kept all other parameters exactly same as suggested in the README commands. Also, if multi-GPU is not supported would you know why the success increases significantly ?

Hello! I'm also trying to change the code to a multi-GPU version using torchrun and deepspeed zero2. But what I get is the checkpoint with weird outputs compared to checkpoints trained with a single GPU, and a strange decrease occurs at the beginning of  training when using multi GPU. But I don't know where the problem is. 

<img width="180" height="234" alt="Image" src="https://github.com/user-attachments/assets/6a6c980b-f934-451e-befd-8b6e260edb20" />

Could you provide how to change the code to multi GPU?

**Comment by zhangquanchen** [comment]
> > Hi,
> > Thanks for the cool work !
> > Does the code currently support multi-GPU training ? I tried training with 8 GPUs using deepspeed zero3 but observed significantly higher success rates for both Stage 1 (+42%) and Stage 2 (+20%) training when comparing to `Avg Success` reported numbers in Fig. 4 of [paper](https://www.arxiv.org/pdf/2506.17218) which I dont think is the expected result. I kept all other parameters exactly same as suggested in the README commands. Also, if multi-GPU is not supported would you know why the success increases significantly ?
> 
> Hello! I'm also trying to change the code to a multi-GPU version using torchrun and deepspeed zero2. But what I get is the checkpoint with weird outputs compared to checkpoints trained with a single GPU, and a strange decrease occurs at the beginning of training when using multi GPU. But I don't know where the problem is.
> 
> <img alt="Image" width="180" height="234" src="https://private-user-images.githubusercontent.com/141009845/483927870-6a6c980b-f934-451e-befd-8b6e260edb20.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY5ODUzOTAsIm5iZiI6MTc1Njk4NTA5MCwicGF0aCI6Ii8xNDEwMDk4NDUvNDgzOTI3ODcwLTZhNmM5ODBiLWY5MzQtNDUxZS1iZWZkLThiNmUyNjBlZGIyMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwNFQxMTI0NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02OTUyNWNjMzM5MzY5ZjE5YjgxNzlmMjBmOTlkYmQxZThkYTU0YzFkZGNmOGE4ZGM5ODE2OGM5ZTNhMTllOWFmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.cNNOS_n5L4KM0V2rc9-bUSD3N7jsyoB_9NrQKe8tULg">
> Could you provide how to change the code to multi GPU?

same request, have you soloved?
